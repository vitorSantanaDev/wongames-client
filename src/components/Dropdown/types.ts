export type DropdownProps = {
  title: React.ReactNode
  children: React.ReactNode
}
