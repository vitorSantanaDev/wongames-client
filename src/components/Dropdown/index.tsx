import { useState } from 'react'
import { DropdownProps } from './types'

import * as S from './styles'

const Dropdown = ({ title, children }: DropdownProps) => {
  const [isOpen, setIsOpen] = useState(false)

  const handleOpenDropdown = () => setIsOpen((prevState) => !prevState)

  return (
    <S.Wrapper isOpen={isOpen}>
      <S.Title onClick={handleOpenDropdown}>{title}</S.Title>
      <S.Content aria-hidden={!isOpen}>{children}</S.Content>
    </S.Wrapper>
  )
}
export default Dropdown
