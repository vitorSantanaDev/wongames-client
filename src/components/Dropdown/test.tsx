import { screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { renderWithTheme } from 'utils/tests/helpers'

import Dropdown from '.'

describe('<Dropdown />', () => {
  beforeEach(() => {
    const title = <h1 aria-label="toggle dropdown">click here</h1>
    renderWithTheme(
      <Dropdown title={title}>
        <span>content</span>
      </Dropdown>
    )
  })

  afterAll(() => {
    const title = <h1 aria-label="toggle dropdown">click here</h1>
    const { container } = renderWithTheme(
      <Dropdown title={title}>
        <span>content</span>
      </Dropdown>
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the title', () => {
    const heading = screen.getByLabelText(/toggle dropdown/)
    expect(heading).toBeInTheDocument()
  })

  it('should handle open/close dropdwon', () => {
    const heading = screen.getByLabelText(/toggle dropdown/)
    const content = screen.getByText(/content/).parentElement

    expect(content).toHaveStyle({ opacity: 0 })
    expect(content?.getAttribute('aria-hidden')).toBe('true')

    userEvent.click(heading)

    expect(content).toHaveStyle({ opacity: 1 })
    expect(content?.getAttribute('aria-hidden')).toBe('false')
  })

  it('should handle open/close dropdwon when clicking on overlay', () => {
    const heading = screen.getByLabelText(/toggle dropdown/)
    const content = screen.getByText(/content/).parentElement
    const overlay = content!.nextElementSibling

    expect(content).toHaveStyle({ opacity: 0 })
    expect(content?.getAttribute('aria-hidden')).toBe('true')

    userEvent.click(heading)

    expect(overlay).toHaveStyle({ opacity: 1 })
    expect(overlay!.getAttribute('aria-hidden')).toBe('false')

    userEvent.click(overlay!)

    expect(overlay).toHaveStyle({ opacity: 0 })
    expect(overlay!.getAttribute('aria-hidden')).toBe('true')
  })
})
