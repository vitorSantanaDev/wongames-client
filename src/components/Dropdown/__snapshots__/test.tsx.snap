// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[
  `<Dropdown /> should handle open/close dropdwon when clicking on overlay 1`
] = `
.c1 {
  cursor: pointer;
  color: #FAFAFA;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-right: 2.4rem;
  z-index: 50;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background: #FAFAFA;
  color: #030517;
  margin-top: 2.4rem;
  position: absolute;
  right: 0;
  z-index: 50;
}

.c3::before {
  content: '';
  position: absolute;
  border-right: 1.2rem solid transparent;
  border-left: 1.2rem solid transparent;
  border-bottom: 1.2rem solid #FAFAFA;
  top: -1.2rem;
  right: 2.4rem;
}

.c5 {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  z-index: 30;
  background-color: rgba(0,0,0,0.5);
}

.c0 {
  position: relative;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
}

.c0 .c2,
.c0 .c4 {
  -webkit-transition: -webkit-transform 0.2s ease-in,opacity 0.3s ease-in-out;
  -webkit-transition: transform 0.2s ease-in,opacity 0.3s ease-in-out;
  transition: transform 0.2s ease-in,opacity 0.3s ease-in-out;
  opacity: 0;
  pointer-events: none;
  -webkit-transform: translateY(-2rem);
  -ms-transform: translateY(-2rem);
  transform: translateY(-2rem);
}

.c6 .c2,
.c6 .c4 {
  -webkit-transition: -webkit-transform 0.2s ease-in,opacity 0.3s ease-in-out;
  -webkit-transition: transform 0.2s ease-in,opacity 0.3s ease-in-out;
  transition: transform 0.2s ease-in,opacity 0.3s ease-in-out;
  opacity: 1;
  pointer-events: auto;
  -webkit-transform: translateY(0);
  -ms-transform: translateY(0);
  transform: translateY(0);
}

<main
  class="c0"
>
  <div
    class="c1"
  >
    <h1
      aria-label="toggle dropdown"
    >
      click here
    </h1>
  </div>
  <div
    aria-hidden="true"
    class="c2 c3"
  >
    <span>
      content
    </span>
  </div>
  <div
    aria-hidden="true"
    class="c4 c5"
  />
</main>
`
